build_native.rs

```diff
@@ -0,0 +1,391 @@
+use std::convert::TryFrom;
+use std::env;
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use anyhow::*;
+use embuild::cargo::IntoWarning;
+use embuild::cmake::codemodel::Language;
+use embuild::cmake::ObjKind;
+use embuild::fs::copy_file_if_different;
+use embuild::git::{CloneOptions, Repository};
+use embuild::python::{check_python_at_least, PYTHON};
+use embuild::utils::{OsStrExt, PathExt};
+use embuild::{bindgen, build, cargo, cmake, cmd, cmd_output, git, kconfig, path_buf};
+use strum::{Display, EnumString};
+
+const SDK_DIR_VAR: &str = "SDK_DIR";
+const ESP_IDF_VERSION_VAR: &str = "ESP_IDF_VERSION";
+const ESP_IDF_REPOSITORY_VAR: &str = "ESP_IDF_REPOSITORY";
+const ESP_IDF_SDKCONFIG_DEFAULTS_VAR: &str = "ESP_IDF_SDKCONFIG_DEFAULTS";
+const ESP_IDF_SDKCONFIG_VAR: &str = "ESP_IDF_SDKCONFIG";
+const ESP_IDF_EXTRA_TOOLS_VAR: &str = "ESP_IDF_EXTRA_TOOLS";
+const MCU_VAR: &str = "MCU";
+
+const DEFAULT_SDK_DIR: &str = ".sdk";
```

Comment:
> What is the reason for keeping the .sdk directory inside the project itself, given that it does not contain anything project-specific (aside from the applied patches, but even for those it can be argued that they are genuinely useful).

My reason was that I don't like that a build script can just install something in my home directory (which is pretty specific to one kind of project). I don't mind it so much for platformio, because if it needs platformio I can guess that the build script needs to install platformio (or just the needed toolchains) for it to work.

I can change the default if you think that makes more sense (and also mention that this build script installs tools in the home directory in the README/docs).

Reply:
I actually don't buy this:
> I don't mind it so much for platformio, because if it needs platformio I can guess that the build script needs to install platformio (or just the needed toolchains) for it to work

If I paraphrase the above for the native build, it would sound just as convincing:
"I don't mind it so much for the Espressif `idf.py` build framework, because if it needs `idf.py` and the ESP-IDF SDK, I can guess that the build script needs to install the Espressif `idf.py` build framework and the ESP-IDF SDK for it to work.

That is to say, whatever we decide to be the default, it should be for *both* build systems. To summarize, perhaps 4 options:
- Option 1 - install and use the default global locations (i.e. `~/.platformio` and `~/.espressif`)
  - Pros: 
    - Does not pollute the binary crate with stuff that should be `.gitignore`-d
    - Downloaded toolchains etc. can be re-used across builds of different binary crates
    - Already downloaded toolchains etc. will just be used
  - Cons:
    - Patching ESP-IDF by the build itself is not so ideal as this is kinda build-specific
    - Messes up with the global user configuration of either PlatformIO or Espressif `idf.py`
- Option 2 - install and use an `embuild`-specific global location (i.e. `~/.embuild/platformio` and `~/.embuild/espressif`)
  - Pros: 
    - Does not pollute the binary crate with stuff that should be `.gitignore`-d
    - Downloaded toolchains etc. can be re-used across builds of different binary crates
    - Already downloaded toolchains etc. will just be used
    - Does not mess up with the user configuration
  - Cons:
    - Patching ESP-IDF by the build itself is not so ideal as this is kinda build-specific
    - Creates something outside the binary crate directory
- Option 3 - install and use an `embuild`-specific location within the *root* of the binary crate (i.e. `.embuild/platformio` and `.embuild/espressif`)
  - Pros: 
    - Does not mess up with the user configuration
    - No global state
    - Patching ESP-IDF OK as it remains build specific 
  - Cons:
    - Pollutes the binary crate with stuff that should be `.gitignore`-d
    - Still does not follow the cargo policy that everything should be in the $OUT_DIR
    - Downloaded toolchains etc. cannot be re-used across builds of different binary crates
- Option 4 - install and use an `embuild`-specific location within the `OUT_DIR` of the binary crate (i.e. `$OUT_DIR/embuild/platformio` and `$OUT_DIR/embuild/espressif`)
  - Pros: 
    - Does not mess up with the user configuration
    - No global state
    - Patching ESP-IDF OK as it remains build specific 
    - Does not pollute the binary crate with stuff that should be `.gitignore`-d
    - Does not follow the cargo policy that everything should be int he $OUT_DIR
    - `cargo clean` means everything is removed, including ESP-IDf patches. Really clean state
  - Cons:
    - `cargo clean` means everything will be re-downloaded again
    - Downloaded toolchains etc. cannot be re-used across builds of different binary crates

Preferences for a default?
Mines are:
* Option 2
* Option 3 (if global state is something you guys feel uneasy) 

In any case, having a flag/switch/env var/whatever so that the build framework is operating in Option 1 mode is a prerequisite IMO.

Reply:
@MabezDev @N3xed @igrr @georgik ?

Reply:
Any feedback? @N3xed - what is your preference?
BTW - I am fine with pushing this, as long as `pio` remains the default in the next week or so. You have commit privileges btw, so you can do this yourself :)